linker/shared/src/main/scala/org/scalajs/linker/checker/IRChecker.scala

```diff
@@ -338,8 +338,8 @@ private final class IRChecker(unit: LinkingUnit, logger: Logger) {
         typecheckExpect(name, Env.empty, AnyType)
     }
 
-    if (fieldDef.ftpe == NoType)
-      reportError(i"FieldDef cannot have type NoType")
+    if (fieldDef.ftpe == NoType || fieldDef.ftpe == NothingType)
+      reportError(i"FieldDef cannot have type ${fieldDef.ftpe}")
```

Comment:
I'm not sure about this one. What I don't like is that we don't have clear guidelines on what invariants to encode as assertions in the Trees, and which ones to check in the IR checker. The only guideline is "if we think it's not going to be a performance problem, put it in the Trees". That's not a very objective criteria.

Since we can't check everything in the Trees for performance reasons, I think we should err on the side of checking less in Trees, and more in the IR checker. Perhaps we should even transfer all the `require`s from `Trees.scala` to checks in the IR checker. At least that way we would have an objective criteria: if we can put it in the (Scala, not IR) type system, it goes to `Trees`. Otherwise it goes to the IR checker.

Reply:
> The only guideline is "if we think it's not going to be a performance problem, put it in the Trees". That's not a very objective criteria.

That's fair. My concern with putting more in the IR checker is that it is much harder to find out how a tree is supposed to be used in the IR checker. Essentially, in terms of ease of use for a developer: type system > assertions > IR checker.

In any case, this is quite independent of this PR and definitely shouldn't block it. I'll review the rest and move this into a separate issue.